src/tribler-core/tribler_core/modules/libtorrent/stream.py

```diff
@@ -1,182 +1,622 @@
+        # or there is already a streaming for a fileindex but we need a new one, reinitialize the
+        # status map. When there is a state for a given infohash, the stream is accepted as streaming,
+        # which means after below line, stream.enaled = True
+        if fileindex != self.fileindex:
+            Stream.status[self.infohash] = [fileindex, [], [], [], {}]
+        elif self.enabled:
+            # if already there is a state with the same file index do nothing
+            if prebufpos is not None:
+                # if the prebuffposiiton is updated, update the static prebuff pieces
+                currrent_prebuf = list(self.prebuffpieces)
+                currrent_prebuf.extend(self.bytestopieces(prebufpos, self.prebuffsize))
+                Stream.status[self.infohash][3] = sorted(list(set(currrent_prebuf)))
+            return
+
+        # update the file name and size with the file index
+        filename, self.filesize = self.files[fileindex]
+        if len(self.files) > 1:
+            self.filename = self.destdir / filename
+        else:
+            self.filename = self.destdir
+        # Backup file prios, and set the the streaming file max prio
+        if not self.enabledfiles:
+            self.enabledfiles = [x[0] for x in enumerate(self.__getfileprios()) if x[1]]
+        self.__setselectedfiles([fileindex], 7, True)
+        # Get the piece map of the file
+        self.firstpiece = self.bytetopiece(0)  # inclusive
+        self.lastpiece = min(self.bytetopiece(self.filesize), len(self.pieceshave) - 1)  # inclusive
+        # prebuffer size PREBUFF_PERCENT of the file size
+        self.prebuffsize = int(self.filesize * PREBUFF_PERCENT)
+        # calculate static buffer pieces
+        headerpieces = self.bytestopieces(0, HEADER_SIZE)
+        footerpieces = self.bytestopieces(-FOOTER_SIZE, 0)
+        prebuffpieces = [] if prebufpos is None else self.bytestopieces(prebufpos, self.prebuffsize)
+        # map pieces to singleton
+        Stream.status[self.infohash][1] = headerpieces
+        Stream.status[self.infohash][2] = footerpieces
+        Stream.status[self.infohash][3] = prebuffpieces
 
-        self.download = download
-        self.info = get_info_from_handle(self.download.handle)
-        self.file = None
-        self.file_size = self.info.file_at(file_index).size
-        self.file_index = file_index
+    @property
+    def enabled(self):
+        """
+        Check if stream is enabled
+        """
+        return self.infohash is not None and self.infohash in Stream.status
 
-        self.prebuffsize = 5 * 1024 * 1024
-        self.endbuffsize = 0
-        self.max_prebuffsize = 5 * 1024 * 1024
-        self.seekpos = 0
+    @property
+    @check_vod(None)
+    def fileindex(self):
+        """
+        Get fileindex of enabled stream, if not enabled returns None
+        """
+        return Stream.status[self.infohash][0]
```

Comment:
Not sure about using this as a singleton (happens in other places too). Can't you just store the status in the `Stream` object?

Reply:
I think you are right, singleton used to make sense when i first started the implementation, but in current state does not make much sense. makes the code look complex than it is.

Reply:
@egbertbouman do you want this change from single to module instance in his PR or another PR is better for this?

Reply:
I think it's OK to do this in this PR.